### PR TITLE
GH-88: System Specs Plugin

### LINKED ISSUE #88: System Specs - Styles & Plugin
# Dependencies

- #83
- #278
- #89
- #298
- #75
- #98
- #310

# Goal

Create a plugin that allows user to enter data for all system specifications pieces at once.

# Requirements

- title
- description
- image
- monitor
- stats data list
- sub-systems description
- sub-systems data list

# Background

In https://github.com/TACC/Core-CMS/milestone/4, the systems page is expected to be created with static styles. Why?
- So styles are created as a component (relies on many classes), not as a scope (relies on a single class and specific markup within).
- Because the dependencies for this task include Sx1 and Mx2 effort tasks, thus this task will take 3 weeks (which is too much to add to Phase 1).

# Reference

- [System Specifications](https://xd.adobe.com/view/f3dad8d1-88ee-40eb-b8da-667b00e4f876-a689/screen/ff0348ff-577d-450a-bef0-5326f3f7c705) (at "System Specifications", "System Image" and the system monitor at the bottom of the image)

### PR DESCRIPTION
**Pull request by tacc-wbomar**
# To Do

- #324
- #278
- #297
- #300
- #308
- #312
- #311
- [x] [Resolve description length inquiry.](https://github.com/TACC/Core-CMS/pull/330#pullrequestreview-752914615)
- [x] [Test suggested boolean logic simplification.](https://github.com/TACC/Core-CMS/pull/330#discussion_r707454767)

# Overview

Create System Specs plugin.

# Issues

- Closes #88

# Screenshots

<details>
<summary>Form</summary>

| Specifications | Footer Link & Subsystems | Sub. Data, Image, Monitor, Adv. |
| --- | --- |  --- |
| GH-88 Form - 1 Top | GH-88 Form - 2 Mid | GH-88 Form - 3 End |

</details>

<details>
<summary>Render</summary>

| 576px | 768px | 992px | 1200px |
| --- |--- | --- | --- |
| GH-88 567px | GH-88 768px | GH-88 992px | GH-88 1200px |

[TACC: Shared UI - Constants - Breakpoints](https://confluence.tacc.utexas.edu/x/b4AZCg)

</details>

<details>
<summary>Structure</summary>

GH-88 Structure

The "Style" elements are a necessary evil until I create Section plugin (GH-91).

</details>

# Testing

1. If no environment, then [prepare to test CMS changes](https://github.com/TACC/Core-CMS/wiki/Test-CMS-Changes).
2. In CMS container, run `python3 manage.py migrate`.
3. On CMS website, logged in as admin, create a new "Frontera System" page.
4. On the new page:

    > Optional steps produce a webpage that more exactly matches the [Frontera System page design].

    - _(optional) Choose "Standard" template (to get Breadcrumbs)._
    - _(optional) Add __two__ "Generic" > "Style" plugins (to create Sections) with:_

        <details>

        - "CLASS NAME": `---------`
        - "TAG TYPE": `section`
        - "Advanced settings" "ADDITIONAL CLASSES": `o-section, o-section--style-light`

        </details>

    - __(required) Add "TACC Site" > "System Specs" plugin (optionally within "Style" plugin).__
    - _(optional) Add "TACC Site" > "Callout" plugin with any content (optionally within "Style" plugin)._
5. Ensure the following behavior:

    <details>
    <summary>Required Behavior</summary>

    1. The "Description" is required.
    2. All "Specifications" specs are optional.
    3. Any "Specifications" specs provided appear as table row with appropriate label and value.
    4. The "Footer link" is optional.
    5. The "Footer link" fields still work like those from a "Generic" > "Link" plugin.
    3. The "Subsystems … - Intro…" "… Title" is optional.
    7. The "Subsystems … - Intro…" "Description" is optional.
    8. The "Subsystems … - Data", "Image", and "System monitor" instructions are clear.
    9. The plugins available for "Subsystems … - Data", "Image", and "System monitor" are:
        - "Generic" > "Image"
        - "TACC Site" > "Data List"
        - "TACC Site" > "System Monitor"
    10. The "Subsystems … - Data", "Image", and "System monitor" instructions work as written:
        - Adding/Nesting a "TACC Site" > "Data List" shows the data on the bottom(-right).
        - Adding/Nesting a "Generic" > "Image" shows the image on the left (or top).
        - Adding/Nesting a "TACC Site" > "System Monitor" shows the monitor a bottom of image.
    11. Styles match those of the [Frontera System page design] with expected differences:
        - Spacing in webpage will be similar to, but more consistent than, Design spacing.
        - Spacing between System Specs and Callout will be 45px (averaged from multiple designs).
        - (I can think of no others. Let me know what you find. I'll confirm whether it's a bug.)
    12. Viewing page on very narrow screen matches [Frontera System page mobile design].

    </details>

# Notes

This plugin will be used first on [Frontera System page], and then on TACC system pages (e.g. [stampede2](https://www.tacc.utexas.edu/systems/stampede2), [longhorn](https://www.tacc.utexas.edu/systems/longhorn), [stockyard](https://www.tacc.utexas.edu/systems/stockyard)).

[Frontera System page design]: https://xd.adobe.com/view/f3dad8d1-88ee-40eb-b8da-667b00e4f876-a689/screen/ff0348ff-577d-450a-bef0-5326f3f7c705

[Frontera System page mobile design]: https://xd.adobe.com/view/f3dad8d1-88ee-40eb-b8da-667b00e4f876-a689/screen/cce23001-6fdd-47fe-bf6b-9e8161f55c85

### CODE REVIEW

File: taccsite_cms/contrib/helpers.py [hunk 1]
```diff
@@ -29,3 +29,41 @@ def concat_classnames(classes):
...
+                    'Only one of ', errors
+                )
+                for i in indices:
+                    err.error_dict[field] = ValidationError(
+                        _('Only one of External link or Internal link may be given.'), code='invalid')
+
+        if len(err.messages) == 0:
+            pass
+        else:
+            raise err
```
Comment (duckonomy) [inline taccsite_cms/contrib/helpers.py:69]: Same suggestion as in the [Callout plugin PR](https://github.com/TACC/Core-CMS/pull/324#discussion_r705520673)
```suggestion
        if len(err.messages):
            raise err
```

**Review by duckonomy — APPROVED** [review]
Everything works well! Will approve.

One question I had was if there should be a limit to how long the description should be?
It seemed to break the style if the description gets too long.

File: taccsite_cms/contrib/helpers.py (hunk 1 above)
Comment (tacc-wbomar) [reply taccsite_cms/contrib/helpers.py]: 👍 

<details><summary>Note to Self</summary>

Self, you tested this change in that PR that duckonomy referenced. Even the single-versus-multi-error concern int he documentation. All seems well.

</details>

**Comment by tacc-wbomar** [comment]
@duckonomy Changes made.

- 755fbbd
- f32ac38


> [Should there] be a limit to how long the description should be?

Yes. Excellent catch. I've set 500, which is:
- twice the description for the first use case (Frontera System Page, in screenshots)
- enough for someone to use more paragraph (but not too much more) than data table
- easy to change if Design later thinks users should be allowed to add more

<details>

The problem with text limits is that on some window widths the text may be "too much" and on some window widths the text is not enough. But some limit should exist to prevent someone from entering the text of a novel, or relying solely on "Description" when they should use a Data List table.

</details>

**Review by duckonomy — APPROVED** [review]
LGTM!

**Pull request merged**